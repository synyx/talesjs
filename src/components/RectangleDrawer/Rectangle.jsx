import React from "react";
import PropTypes from "prop-types";

export default class Rectangle extends React.Component {
  static propTypes = {
    x: PropTypes.number.isRequired,
    y: PropTypes.number.isRequired,
    width: PropTypes.number.isRequired,
    height: PropTypes.number.isRequired,
    onMove: PropTypes.func,
    onMoveEnd: PropTypes.func,
  };

  static defaultProps = {
    onMove: () => {},
    onMoveEnd: () => {},
  };

  state = {
    selected: false,
  };

  handleRectClick(event) {
    event.stopPropagation();
    this.setState({ selected: !this.state.selected });
  }

  handleMouseDown(event) {
    this.start = {
      x: event.clientX,
      y: event.clientY,
    };
  }

  handleMouseUp(event) {
    this.start = null;
    this.props.onMoveEnd();
  }

  handleMouseMove(event) {
    if (this.start) {
      event.stopPropagation();
      this.props.onMove({
        dx: event.clientX - this.start.x,
        dy: event.clientY - this.start.y,
      });
    }
  }

  render() {
    const { x, y, width, height } = this.props;
    const { selected } = this.state;
    return (
      <rect
        x={x}
        y={y}
        width={width}
        height={height}
        onClick={event => this.handleRectClick(event)}
        stroke={selected ? "purple" : "blue"}
        fill="rgba(0,0,0,0.8)"
        onMouseDown={event => this.handleMouseDown(event)}
        onMouseUp={event => this.handleMouseUp(event)}
        onMouseMove={event => this.handleMouseMove(event)}
      />
    );
  }
}
