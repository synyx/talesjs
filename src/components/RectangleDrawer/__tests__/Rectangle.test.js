import React from "react";
import Rectangle from "../Rectangle";
import { renderShallow } from "../../../../test";

describe("Rectangle", () => {
  it("renders", () => {
    const props = setupProperties();
    const tree = renderShallow(<Rectangle {...props} />);
    expect(tree).toMatchSnapshot();
  });

  function setupProperties() {
    return {
      x: 10,
      y: 20,
      width: 800,
      height: 600,
    };
  }
});
