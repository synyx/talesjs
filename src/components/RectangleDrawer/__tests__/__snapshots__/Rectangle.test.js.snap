// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rectangle renders 1`] = `
<g
  onMouseDown={[Function]}
>
  <rect
    fill="rgba(0,0,0,0.8)"
    height={600}
    onClick={[Function]}
    stroke="blue"
    width={800}
    x={10}
    y={20}
  />
  <rect
    fill="rgba(0,0,0,0.5)"
    height={10}
    stroke="orange"
    style={
      Object {
        "cursor": "ns-resize",
      }
    }
    width={10}
    x={410}
    y={15}
  />
  <rect
    fill="rgba(0,0,0,0.5)"
    height={10}
    stroke="orange"
    style={
      Object {
        "cursor": "ns-resize",
      }
    }
    width={10}
    x={410}
    y={615}
  />
  <rect
    fill="rgba(0,0,0,0.5)"
    height={10}
    stroke="orange"
    style={
      Object {
        "cursor": "ew-resize",
      }
    }
    width={10}
    x={5}
    y={320}
  />
  <rect
    fill="rgba(0,0,0,0.5)"
    height={10}
    stroke="orange"
    style={
      Object {
        "cursor": "ew-resize",
      }
    }
    width={10}
    x={805}
    y={320}
  />
</g>
`;
